nimsuggest/nimsuggest.nim

```diff
@@ -647,6 +666,164 @@ proc handleCmdLine(cache: IdentCache; conf: ConfigRef) =
+      line: toLinenumber(info), column: toColumn(info), doc: msg, forth: $sev)
+    graph.suggestErrors.mgetOrPut(info.fileIndex, @[]).add suggest
+  var isKnownFile = true
+  let conf = graph.config
+  conf.ideCmd = cmd
+
+  myLog fmt "cmd: {cmd}, file: {file}[{line}:{col}], dirtyFile: {dirtyfile}"
+
+  msgs.setDirtyFile(
+    conf,
+    fileInfoIdx(conf, file),
+    if dirtyfile.isEmpty: AbsoluteFile"" else: dirtyfile)
+
+  if not dirtyfile.isEmpty:
+    graph.markDirtyIfNeeded(dirtyFile.string, fileInfoIdx(conf, file))
+
+  # these commands require fully compiled project
+  if cmd in {ideUse, ideDus, ideGlobalSymbols, ideChk} and graph.needsCompilation():
+    graph.recompilePartially()
+    # when doing incremental build for the project root we should make sure that
+    # everything is unmarked as no longer beeing dirty in case there is no
+    # longer reference to a particular module. E. g. A depends on B, B is marked
+    # as dirty and A loses B import.
+    graph.unmarkAllDirty()
+
+  # these commands require partially compiled project
+  # TODO: we should should check only if there are downstream
+  elif cmd in {ideSug, ideOutline, ideHighlight, ideDef}:
+    let fileIndex = fileInfoIdx(conf, file)
+    if graph.needsCompilation(fileIndex):
+      graph.recompilePartially(fileIndex)
+
+  case cmd
+  of ideDef:
+    let
+      fileIdx = fileInfoIdx(conf, file)
+      (sym, info) = graph.findSymData(file, line, col)
+    if sym != nil:
+      graph.suggestResultUseOrDef(sym, info)
+  of ideUse, ideDus:
+    let symbol = graph.findSymData(file, line, col).sym
+    if symbol != nil:
+      for (sym, info) in graph.suggestSymbolsIter:
+        if sym == symbol:
+          graph.suggestResultUseOrDef(sym, info)
+  of ideHighlight:
+    let sym = graph.findSymData(file, line, col).sym
+    if sym != nil:
+      let usages = graph
+        .suggestSymbols
+        .getOrDefault(fileInfoIdx(conf, file))
+        .filterIt(it.sym == sym)
+      myLog fmt "Found {usages.len} usages in {file.string}"
+      for (sym, info) in usages:
+        graph.suggestResultUseOrDef(sym, info)
+  of ideRecompile:
+    if file.string != "clean": graph.recompileFullProject_v3()
+    else: graph.recompilePartially()
+  of ideChanged:
+    graph.markDirtyIfNeeded(file.string, fileInfoIdx(conf, file))
+  of ideOutline:
+    let
+      fileIdx = fileInfoIdx(conf, file)
+      module = graph.getModule fileIdx
+      symbols = graph.suggestSymbols
+        .getOrDefault(fileIdx)
+        .filterIt(it.sym.info == it.info and it.sym.owner == module)
+    for (sym, _) in symbols:
+      suggestResult(
+        conf,
+        symToSuggest(graph, sym, false,
+                     ideOutline, sym.info, 100, PrefixMatch.None, false, 0))
+  of ideChk:
+    myLog fmt "Reporting {graph.suggestErrors.len} error(s)"
+    for sug in graph.suggestErrorsIter:
+      suggestResult(graph.config, sug)
+  of ideGlobalSymbols:
+    var counter = 0
+    let reg = re(string(file))
```

Comment:
Is there a need for this to be a regex? Can't it be done with contains (`in`) or something similar?

Reply:
Code below seems to do `find(cstring(sym.name.s), reg, 0, sym.name.s.len)` , you can instead just use https://nim-lang.org/docs/strutils.html#rfind%2Cstring%2Cstring%2CNatural%2Cint

Reply:
@yyoncho the problem is that it affects more than just personal preference - Nim's `re` library depends on PCRE, so if you make nimsuggest depend on re, users will have to make sure they have pcre installed. In a simple case like this I really think you should just use strutils.find instead.

Reply:
> @yyoncho the problem is that it affects more than just personal preference

Of course - I am just saying why I picked this implementation. Using `find` should be good enough for the initial implementation.

Reply:
@Yardanico changed to contains. I was wondering whether to downcase before matching but I guess it might lead to too many matches.